source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -60,6 +60,17 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
   ASSERT(!callbacks_);
   callbacks_ = &callbacks;
 
+  // Associate this SSL connection with all the certificates (with their potentially different
```

Comment:
The private key method is going to be called only on one of the configured providers, so registering it on all is a bit wasteful. Could we auto-register on `Sign()` / `Decrypt()` operations instead, somehow?

Reply:
We could maybe add a pointer to the `SslSocket` into the SSL object using `SSL_set_ex_data()` during SSL object creation, and then add a public function to `SslSocket` to get access to the thread `Dispatcher` object. Then that `SslSocket` pointer could be retrieved in `sign()`/`decrypt()` and `complete()` functions to get access to the dispatcher and to the callback object. Since there would be no pre-registration of provider to SSL object, we would probably need to add the private key method provider pointer to `SSL_CTX` object using `SSL_CTX_set_ex_data()`, so that provider configuration could be accessed from `sign()`/`decrypt()`.

However, there are some issues with this approach. Because we can't use references or smart pointers in SSL object user data functions, we will be reduced to using plain void pointers and casting. Following the program logic might become bit somewhat more complex too. Also, the maximum amount of certificates supported for a TLS context in the multi-cert use case is two, so we are dealing with small constants instead of an arbitrary number of certificates. Still, I can do the change if you believe this would be a good idea. What do you think?

Reply:
Let's leave it as-is. Thanks.